lib/asciidoctor/reader.rb

```diff
@@ -129,16 +129,16 @@ def next_line_empty?
   # Returns the next line of the source data as a String if there are lines remaining.
   # Returns nothing if there is no more data.
   def peek_line direct = false
-    if direct || @look_ahead > 0
-      @unescape_next_line ? ((line = @lines[-1]).slice 1, line.length) : @lines[-1]
-    elsif @lines.empty?
-      @look_ahead = 0
-      nil
-    else
+    while true
+      return @unescape_next_line ? ((line = @lines[-1]).slice 1, line.length) : @lines[-1] if direct || @look_ahead > 0
+      if @lines.empty?
+        @look_ahead = 0
+        return
+      end
       # FIXME the problem with this approach is that we aren't
       # retaining the modified line (hence the @unescape_next_line tweak)
       # perhaps we need a stack of proxied lines
-      (process_line @lines[-1]) || peek_line
+      return (process_line @lines[-1]) || redo
```

Comment:
It seems to be supported in Opal, not sure if the implementation is performant or not, I will give it a try.

Reply:
I found a slightly more portable way of doing it that doesn't rely on redo and seems to be even faster.

Reply:
Perfect 👌🏻